src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs

```diff
@@ -413,6 +413,9 @@ private HttpMessageHandlerStage SetupHandlerChain()
                 throw new NotSupportedException(SR.Format(SR.net_http_http2_sync_not_supported, GetType()));
             }
 
+            // Do not allow upgrades for synchronous requests, that might lead to asynchronous code-paths.
+            request.VersionPolicy = HttpVersionPolicy.RequestVersionOrLower;
```

Comment:
Doesn't this mean someone could specify RequestVersionOrHigher, and they may silently get a lower version?  That seems wrong.

Reply:
Fixed, now we throw when user requests upgrades.